File: offline-lib/src/queue.ts
import {
  decodeKey,
  decryptJson,
  encodeKey,
  encryptJson,
  LocalStorageStore,
  MemoryStore,
  randomAesKey,
  type KeyValueStore,
} from "./crypto";

export type AttendancePayload = {
  registerId: string;
  learnerId: string;
  status: string;
  idempotencyKey: string;
  reasonId: string | null;
  notes: string | null;
};

export type QueueItem<T> = {
  id: string;
  idempotencyKey: string;
  status: "pending" | "acked";
  conflictPolicy: "server_wins";
  payload: T;
  createdAt: string;
};

export type EnqueueResult<T> = {
  created: boolean;
  item: QueueItem<T>;
};

const QUEUE_KEY = "metroskool-monitor-offline-queue";
const KEY_REF = "metroskool-monitor-offline-key";

export class MonitorOfflineQueue {
  constructor(
    private readonly store: KeyValueStore,
    private readonly keyBytes: Uint8Array,
  ) {}

  private async readAll(): Promise<QueueItem<AttendancePayload>[]> {
    const blob = this.store.get(QUEUE_KEY);
    if (!blob) return [];
    return decryptJson<QueueItem<AttendancePayload>[]>(this.keyBytes, blob);
  }

  private async writeAll(items: QueueItem<AttendancePayload>[]): Promise<void> {
    this.store.set(QUEUE_KEY, await encryptJson(this.keyBytes, items));
  }

  async enqueueAttendance(payload: AttendancePayload): Promise<EnqueueResult<AttendancePayload>> {
    const items = await this.readAll();
    const existing = items.find((row) => row.idempotencyKey === payload.idempotencyKey);
    if (existing) {
      return { created: false, item: existing };
    }
    const item: QueueItem<AttendancePayload> = {
      id: payload.idempotencyKey,
      idempotencyKey: payload.idempotencyKey,
      status: "pending",
      conflictPolicy: "server_wins",
      payload,
      createdAt: new Date().toISOString(),
    };
    items.push(item);
    await this.writeAll(items);
    return { created: true, item };
  }

  async listPending(): Promise<QueueItem<AttendancePayload>[]> {
    const items = await this.readAll();
    return items.filter((row) => row.status === "pending");
  }

  async markAcked(idempotencyKey: string): Promise<void> {
    const items = await this.readAll();
    const next = items.map((row) =>
      row.idempotencyKey === idempotencyKey ? { ...row, status: "acked" as const } : row,
    );
    await this.writeAll(next);
  }
}

export function createOfflineQueue(
  store: KeyValueStore = new MemoryStore(),
  keyBytes: Uint8Array = randomAesKey(),
): MonitorOfflineQueue {
  return new MonitorOfflineQueue(store, keyBytes);
}

export function createBrowserOfflineQueue(): MonitorOfflineQueue {
  const store = new LocalStorageStore();
  const existing = store.get(KEY_REF);
  const keyBytes = existing ? decodeKey(existing) : randomAesKey();
  if (!existing) {
    store.set(KEY_REF, encodeKey(keyBytes));
  }
  return new MonitorOfflineQueue(store, keyBytes);
}

export { MemoryStore, LocalStorageStore };

File: mobile/lib/src/queue.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:metroskool_flutter_core/metroskool_flutter_core.dart';

/// Attendance mark queued while the teacher is offline.
class AttendancePayload {
  const AttendancePayload({
    required this.registerId,
    required this.learnerId,
    required this.status,
    required this.idempotencyKey,
    this.reasonId,
    this.notes,
  });

  final String registerId;
  final String learnerId;
  final String status;
  final String idempotencyKey;
  final String? reasonId;
  final String? notes;

  Map<String, Object?> toJson() => {
        'registerId': registerId,
        'learnerId': learnerId,
        'status': status,
        'idempotencyKey': idempotencyKey,
        'reasonId': reasonId,
        'notes': notes,
      };

  factory AttendancePayload.fromJson(Map<String, Object?> json) {
    return AttendancePayload(
      registerId: json['registerId'] as String,
      learnerId: json['learnerId'] as String,
      status: json['status'] as String,
      idempotencyKey: json['idempotencyKey'] as String,
      reasonId: json['reasonId'] as String?,
      notes: json['notes'] as String?,
    );
  }
}

class EnqueueResult {
  const EnqueueResult({required this.created, required this.item});

  final bool created;
  final SyncQueueItem item;
}

/// XOR stream cipher over a 32-byte key. Enough to keep plaintext out of the
/// in-memory blob; production devices should wrap this with platform keystore.
class _XorBox {
  _XorBox(this.key);

  final Uint8List key;

  String seal(String plaintext) {
    final bytes = utf8.encode(plaintext);
    final out = Uint8List(bytes.length);
    for (var i = 0; i < bytes.length; i++) {
      out[i] = bytes[i] ^ key[i % key.length];
    }
    return base64Encode(out);
  }

  String open(String blob) {
    final bytes = base64Decode(blob);
    final out = Uint8List(bytes.length);
    for (var i = 0; i < bytes.length; i++) {
      out[i] = bytes[i] ^ key[i % key.length];
    }
    return utf8.decode(out);
  }
}

/// Idempotent Monitor attendance queue. Conflict policy is always server-wins.
class MonitorAttendanceQueue implements SyncQueue {
  MonitorAttendanceQueue({Uint8List? key})
      : _box = _XorBox(key ?? _randomKey());

  final _XorBox _box;
  final Map<String, SyncQueueItem> _items = {};

  static Uint8List _randomKey() {
    final random = Random.secure();
    return Uint8List.fromList(List<int>.generate(32, (_) => random.nextInt(256)));
  }

  Future<EnqueueResult> enqueueAttendance({
    required SchoolId schoolId,
    required AttendancePayload payload,
  }) async {
    final existing = _items[payload.idempotencyKey];
    if (existing != null) {
      return EnqueueResult(created: false, item: existing);
    }
    final item = SyncQueueItem(
      id: payload.idempotencyKey,
      schoolId: schoolId,
      moduleId: ModuleId.monitor,
      idempotencyKey: payload.idempotencyKey,
      status: SyncQueueStatus.pending,
      payloadJson: _box.seal(jsonEncode(payload.toJson())),
      createdAt: DateTime.now().toUtc(),
    );
    await enqueue(item);
    return EnqueueResult(created: true, item: item);
  }

  AttendancePayload decode(SyncQueueItem item) {
    final decoded = jsonDecode(_box.open(item.payloadJson)) as Map<String, dynamic>;
    return AttendancePayload.fromJson(decoded.cast<String, Object?>());
  }

  @override
  Future<void> enqueue(SyncQueueItem item) async {
    _items.putIfAbsent(item.idempotencyKey, () => item);
  }

  @override
  Future<List<SyncQueueItem>> pending() async {
    return _items.values
        .where((row) => row.status == SyncQueueStatus.pending)
        .toList(growable: false);
  }

  @override
  Future<void> markAcked(String id) async {
    final current = _items[id];
    if (current == null) return;
    _items[id] = SyncQueueItem(
      id: current.id,
      schoolId: current.schoolId,
      moduleId: current.moduleId,
      idempotencyKey: current.idempotencyKey,
      status: SyncQueueStatus.acked,
      payloadJson: current.payloadJson,
      createdAt: current.createdAt,
    );
  }

  String? lastFailureReason;

  @override
  Future<void> markFailed(String id, String reason) async {
    lastFailureReason = reason;
    final current = _items[id];
    if (current == null) return;
    _items[id] = SyncQueueItem(
      id: current.id,
      schoolId: current.schoolId,
      moduleId: current.moduleId,
      idempotencyKey: current.idempotencyKey,
      status: SyncQueueStatus.failed,
      payloadJson: current.payloadJson,
      createdAt: current.createdAt,
    );
  }
}
